### PR TITLE
Refactors Sockets closed promise handling.

### PR DESCRIPTION
**Pull request by dom96**
This is a more basic version of https://github.com/cloudflare/workerd/pull/205. The main change in semantics here is that the `closed` promise will now have any errors propagated to it (from the readable/writable stream close calls). The rest is a major refactor to reuse the existing `newPromisedStream`.

### Test Plan

```
$ bazel test //src/ew/tests:sockets/sockets-tcp-test.ew-test-bin --config=asan --test_timeout 5
```

### CODE REVIEW

File: src/workerd/api/system-streams.c++
Comment (harrishancock) [inline src/workerd/api/system-streams.c++:339]: I'm not sure the `abortRead()` makes sense (unless that's what the Encoded classes do?). IIRC, the Streams spec uses the word abort on the writable side of things, while KJ uses abortRead on the readable side of things.

File: src/workerd/api/system-streams.c++
Comment (harrishancock) [inline src/workerd/api/system-streams.c++:330]: I think this will recurse forver -- check out the definition of `ReadableStreamSource::pumpTo()`: https://github.com/cloudflare/workerd/blob/main/src/workerd/api/streams/internal.c%2B%2B#L333

Instead, this is our opportunity to optimize the pump. We can try downcasting `ReadableStreamSource` to a `WrappedAsyncIoStream` (see the first line of `EncodedAsyncOutputStream::tryPumpFrom()` for an example) and, if that returns non-null, this optimizes to `wrappedInput->inner->pumpTo(*inner)`, and we can even return a deferred proxy promise.

If `input` is _not_ a `WrappedAsyncIoStream`, then the correct thing to do is return nullptr, which causes the calling `pumpTo()` to fall back to its naive implementation.

If you just want to get things going, I think it'd be correct to unconditionally return nullptr, just suboptimal.

File: src/workerd/api/system-streams.c++
Comment (harrishancock) [inline src/workerd/api/system-streams.c++:304]: If the output encoding is not identity, I think we should just return nullptr. This function is saying, "I plan to read everything from this stream and immediately encode it in `outEncoding`. How many bytes is that going to be?"

File: src/workerd/api/sockets.c++
Comment (harrishancock) [inline src/workerd/api/sockets.c++:56]: We'll be registering pending events on this stream's I/O operations. In the period of time where `connectionPromise` hasn't yet resolved, that means we'll be registering pending events to keep alive `connectionPromise`. I _think_ we can assume that `connectionPromise` is backed by system I/O, so that this is safe. I'd like @kentonv's opinion too, though.

File: src/workerd/api/sockets.h
Comment (harrishancock) [inline src/workerd/api/sockets.h:21]: I don't think this needs to be wrapped in an IoOwn, now that it's a jsg type.

File: src/workerd/api/sockets.h
Comment (harrishancock) [inline src/workerd/api/sockets.h:54]: No need for IoOwn here anymore.

File: src/workerd/api/sockets.c++
Comment (harrishancock) [inline src/workerd/api/sockets.c++:74]: Is it okay if two calls to this function back-to-back both call cancel() and abort()? That is, should we have a concept of a "closing" state?

File: src/workerd/api/system-streams.c++
Comment (kentonv) [inline src/workerd/api/system-streams.c++:334]: I think you want to do `inner->shutdownWrite()` here.

File: src/workerd/api/system-streams.c++
Comment (kentonv) [reply src/workerd/api/system-streams.c++:330]: Oh, this seems like a good argument not to introduce a new type at all, and instead try to reuse `EncodedAsync*Stream`. What if you are pumping from an `EncodedAsyncInputStream` to a `WrappedAsyncIoStream`? This pump should be optimized, too, as should `WrappedAsyncIoStream` -> `EncodedAsyncOutputStream`.

If we keep using the encoded streams then this all just works without quadratic checks. Setting the encoding to `IDENTITY` seems fine.

The only thing is, we need `EncodedAsyncOutputStream::end()` to call `shutdownWrite()`, which is awkward because it doesn't know that the inner stream is an `AsyncIoStream`.

Actually, is this buggy today? It's weird that `end()` does not indicate EOF to the underlying stream at all, unless the `EncodedAsyncOutputStream` itself is dropped immediately after...

File: src/workerd/api/sockets.c++
Comment (kentonv) [reply src/workerd/api/sockets.c++:56]: Does `connectionPromise` even need to be a promise anymore, now that the underlying connect API supports pipelining?

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:74]: Good question. I will add a TODO comment here to add a test for this.

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++]: I was really puzzled by how I was still getting a Promise from HttpClient::connect, turns out the `.attach` caused the `Own` to be turned into a `Promise`. Not something I would have expected. But anyway, resolved.

**Comment by dom96** [comment]
Last push was a rebase.

Read for another review

File: src/workerd/api/system-streams.c++
Comment (kentonv) [reply src/workerd/api/system-streams.c++]: @dom96 This was marked resolved but it looks like while Harris's comment was addressed, mine was not.

File: src/workerd/api/sockets.c++
Comment (kentonv) [inline src/workerd/api/sockets.c++:46]: Obsolete comment?

**Comment by kentonv** [comment]
Please note my reply to a previous thread (which had been marked resolved): https://github.com/cloudflare/workerd/pull/237#discussion_r1055815436

File: src/workerd/api/system-streams.c++
Comment (dom96) [reply src/workerd/api/system-streams.c++]: @kentonv Can you clarify which part?

I do reuse the `EncodedAsync*Stream` here: https://github.com/cloudflare/workerd/pull/237/files#diff-5f332a1e13274be21437d308032286aa3197c8158f8a9274c814296b3147691fR298 and I do call `shutdownWrite` here: https://github.com/cloudflare/workerd/pull/237/files#diff-5f332a1e13274be21437d308032286aa3197c8158f8a9274c814296b3147691fR336.

**Comment by dom96** [comment]
@kentonv responded.

File: src/workerd/api/system-streams.c++
Comment (dom96) [reply src/workerd/api/system-streams.c++]: @harrishancock appreciate you taking a look as well since we discussed how to solve this, would like to make sure I didn't miss anything in my implementation

Comment (kentonv) [reply src/workerd/api/system-streams.c++]: Oh I see, I missed the fact that `EncodedAsyncIoStream` is now a wrapper around a pair of `EncodedAsyncInputStream` and `EncodedAsyncOutputStream`.

But why? `newSystemMultiStream()` itself returns a pair of objects `readable` and `writable`, but as currently implemented they are both refcounts of a single `EncodedAsyncIoStream`, which in turn wraps a pair of `EncodedAsyncInputStream` and `EncodedAsyncOutputStream`.

Why not have `newSystemMultiStream()` just return the separate `EncodedAsyncInputStream` and `EncodedAsyncOutputStream`, and delete the `EncodedAsyncIoStream` class?

Comment (dom96) [reply src/workerd/api/system-streams.c++]: IIRC I need the wrapper `EncodedAsyncIoStream` class to be able to call `shutdownWrite` on the `AsyncIoStream` when `end()` is called.

Comment (kentonv) [reply src/workerd/api/system-streams.c++]: I don't think you do. `EncodedAsyncOutputStream::end()` can call `shutdownWrite()`. A couple possible approaches:

* It could use `dynamic_cast` to test if the inner stream implements `AsyncIoStream` and, if so, call `shutdownWrite()`.
* It could have an additional constructor which accept `AsyncIoStream`, and it could store multiple references -- `kj::Own<kj::AsyncInputStream> inner` and `kj::Maybe<kj::AsyncIoStream&> ioInner` -- the latter is only non-null if the input was AsyncIoStream.

The latter approach is somewhat cleaner but also more involved. I would be fine with either.

Incidentally in the case that the inner object is _not_ `AsyncIoStream`, but just regular `AsyncInputStream`, then the way to signal EOF is to destroy the object. But, `end()` isn't do that today! This seems like a bug. I suppose it only works because the owner of the `WritableStreamSink` always drops the object immediately after `end()` finishes?

**Comment by dom96** [comment]
Rebased

File: src/workerd/api/system-streams.c++
Comment (dom96) [reply src/workerd/api/system-streams.c++]: hm, okay, I updated to use `dynamic_cast` though I think I am missing why this approach is significantly better than what I had. Is it mainly the reduction in indirection?

I didn't add any further destruction of the inner object if it's not an `AsyncIoStream`, if you think I should add that can you let me know what the best way to destroy the object explicitly there is?

**Comment by dom96** [comment]
Ready for review.

**Review by jasnell — APPROVED** [review]
LGTM but let's also have @kentonv or @harrishancock review and sign off.

File: src/workerd/api/system-streams.c++
Comment (kentonv) [reply src/workerd/api/system-streams.c++]: Well, I had been hoping you could get rid of the stream wrapper altogether, but I see you still have it just for the sake of refcounting. Maybe you could do the same trick we did in the http-over-capnp PR?

```suggestion
  struct RefcountedIoStream: public kj::Refcounted {
    kj::Own<AsyncIoStream> inner;
    RefcountedIoStream(kj::Own<AsyncIoStream> inner): inner(kj::mv(inner)) {}
  };
  auto rc = kj::refcounted<RefcountedIoStream>(kj::mv(stream));
  kj::Own<kj::AsyncIoStream> ref1 = kj::attachRef(*inner.get(), kj::addRef(*rc));
  kj::Own<kj::AsyncIoStream> ref2 = kj::attachRef(*inner.get(), kj::mv(*rc));
```

(Perhaps we should add this pattern to KJ...)

Comment (kentonv) [reply src/workerd/api/system-streams.c++]: I made a KJ PR to formalize this pattern: https://github.com/capnproto/capnproto/pull/1605

Comment (dom96) [reply src/workerd/api/system-streams.c++]: I spent some time trying to figure out how to do this again and completely forgot that used this pattern in the http-over-capnp PR. Anyway, thanks for creating that PR, I'll wait until it gets merged and modify this PR to use it.

Comment (kentonv) [reply src/workerd/api/system-streams.c++]: Merged. Don't forget we'll need to update the capnproto submodule on the internal repo as well. (Something I myself forgot to do recently with a workerd PR...)

File: src/workerd/api/system-streams.c++
Comment (jasnell) [inline src/workerd/api/system-streams.c++:261]: This can be replaced with the new `kj::refcountedWrapper(...)` correct?

Comment (kentonv) [reply src/workerd/api/system-streams.c++:261]: That's the plan, see thread above: https://github.com/cloudflare/workerd/pull/237#discussion_r1066867895

**Comment by dom96** [comment]
Rebased

File: src/workerd/api/sockets.c++
Comment (kentonv) [inline src/workerd/api/sockets.c++:81]: Why `context.addTask(context.run(`? This seems redundant, but forces exiting and re-entering the isolate. Can you just execute the code below inline?

File: src/workerd/api/sockets.c++
Comment (kentonv) [inline src/workerd/api/sockets.c++:84]: Any particular reason `redable.cancel()` and `writable.abort()` have to be in series rather than parallel?

File: src/workerd/api/sockets.c++
Comment (kentonv) [inline src/workerd/api/sockets.c++:75]: This `isClosed` check looks race-y since `isClosed` doesn't actually become true until after the async `cancel()` and `abort()` calls below complete. So if `close()` is called twice in a row, both calls will continue on to the `cancel()` and `abort()` below. If the check isn't really important than I'd remove it, but if it is important then it currently doesn't look like it works.

File: src/workerd/api/sockets.h
Comment (kentonv) [inline src/workerd/api/sockets.h:69]: Unused?

File: src/workerd/api/sockets.h
Comment (kentonv) [inline src/workerd/api/sockets.h:30]: It looks like, at present, this promise is only resolved when the application calls `close()` explicitly -- not when the remote side disconnects. Is that correct?

If so, it seems like it's not accomplishing a whole lot to offer a way to listen for closure. The application can set that up itself and save us the trouble.

If the intent is that this promise should also tell the application when the remote end disconnects the socket, then I think you should add code that listens on `whenWriteDisconnected()` and fulfills the promise at that point. (Or maybe it should wait on both `whenWriteDisconnected()` and EOF or error on the read side, and only resolve when both are true... I'm not really sure... I'd be satisfied with waiting for `whenWriteDisconnected()` only.)

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:81]: Here is why I ended up with this, it's possible there is another way to achieve each of these steps differently to avoid doing what I'm doing here:

* In order to call `cancel` and `abort` I need the `jsg::Lock`
* In order to get the `jsg::Lock` I need to call `context.run`
* The method doesn't return a Promise, so `context.addTask` on that promise so it doesn't get lost

As a side note, I think we will end up with `close` returning a Promise again. But we need to discuss this aspect of the API and make a decision so I'd rather leave it as-is in this PR.

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:75]: Yeah, that's why I have the "todo" in there :)

This check will still help if `close` is called once the `closed` promise resolves by accident for example.

I'd rather keep it and iterate on it to resolve the race later./

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:84]: No reason. I just figured it was marginally better to try cancelling one of them and not bother cancelling the other if the first cancel operation fails. Any reason doing this in parallel would be better?

File: src/workerd/api/sockets.h
Comment (dom96) [reply src/workerd/api/sockets.h:69]: yep, good catch.

File: src/workerd/api/sockets.h
Comment (dom96) [reply src/workerd/api/sockets.h:30]: Yep, it's the current behaviour in workerd HEAD as well. I split this PR off from https://github.com/cloudflare/workerd/pull/205 to not block on this and instead focus on the refactoring.

I will follow up on this but first we need to decide how we want the semantics to work exactly.

File: src/workerd/api/sockets.c++
Comment (kentonv) [reply src/workerd/api/sockets.c++:81]: You can add a parameter `jsg::Lock& js` to `close()` itself, and JSG will automatically fill it in.

It might be a good idea to re-read the JSG docs.

File: src/workerd/api/sockets.c++
Comment (kentonv) [reply src/workerd/api/sockets.c++:84]: Well, it seems to me that after `close()` is called, we'd want any methods called on `writable` to immediately behave as if `abort()` had been called on it. But with this implementation, `abort()` doesn't actually get called until a while later.

That said, it's not clear to me why `cancel()` and `abort()` aren't synchronous. When would these calls ever need to wait for I/O, just to cancel it? I guess that's a question for the streams authors, though.

Comment (kentonv) [reply src/workerd/api/sockets.c++:84]: Relatedly, I feel like `close()` should not return a promise, but for the same reason I think `cancel()` and `abort()` shouldn't. But `cancel()` and `abort()` _do_ return promises, so perhaps `close()` should as well. I suppose it should return essentially `Promise.all(readable.cancel(), writable.abort())`.

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:81]: Ahh. I knew about this pattern but it didn't click in my head that `close` is a JSG_METHOD that can use it...

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:84]: Hm, actually now that I tried to refactor this I realised the real reason I did it this way: it's easier to resolve the `closed` promise successfully once I know that both readable and writable have been closed. If I call them in parallel then it becomes tricky, because when do I resolve the promise? I need to know when both have been closed successfully so I might as well close the synchronously. Note that the `cancel` and `abort` operations can fail, and their failure causes the `closed` promise to be resolved with an error.

Hope that makes sense and I didn't miss anything.

File: src/workerd/api/sockets.c++
Comment (jasnell) [reply src/workerd/api/sockets.c++:75]: I would add a "closing" state.

File: src/workerd/api/sockets.c++
Comment (jasnell) [reply src/workerd/api/sockets.c++]: Returning a Promise.all from close should work, I'd think?

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:75]: Turns out cancelling/aborting the streams multiple times doesn't create any problems, so I ended up just removing these states completely. Upstream I've added a test to ensure this case is covered as well.

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++]: Chatted with James about this: TL;DR added a TODO for this.

File: src/workerd/api/sockets.h
Comment (jasnell) [inline src/workerd/api/sockets.h:24]: One other bit here. Currently, the way this is set up, if the `closedPromise` rejects and there is no one handling rejections on the `socket.closed` property, an `unhandledrejection` event will occur. I'm not convinced that's ideal by default. We should probably mark the `closedPromise` as handled explicitly so that the unhandled rejection event only occurs if someone attaches a `then` handler to the promise without a corresponding `catch` handler.

File: src/workerd/api/sockets.c++
Comment (jasnell) [inline src/workerd/api/sockets.c++:82]: I *believe* you only need the catch handler on the outer promise here.

Comment (kentonv) [inline src/workerd/api/sockets.c++:82]: So it looks like any exceptions caught from `cancel()` or `abort()` propagate to the `closed` promise, but do NOT propagate to the result of `close()` itself.

Is that what we want? I'm not really sure when `cancel()` or `abort()` would throw, but it would seem to me that the exception should propagate out of `close()`. Meanwhile it's not clear to me whether the `closed` promise should signal errors or just signal that the socket is now closed.

File: src/workerd/api/sockets.h
Comment (kentonv) [inline src/workerd/api/sockets.h:55]: I'd suggest making `jsg::Lock&` be a parameter to `resolveFulfiller()` so that you don't have to look up the current `IoContext`. Both call sites have a `jsg::Lock` readily available.

`context.getCurrentLock()` is kind of a hack that I'd like to remove eventually.

File: src/workerd/api/sockets.h
Comment (dom96) [reply src/workerd/api/sockets.h:24]: I suppose now that we have `close` returning a promise again we can instead reject that returned promise instead of the `closed` promise (and I see below that this is what Kenton is suggesting we should do). So I'll do that and that will resolve this issue too.

Comment (dom96) [reply src/workerd/api/sockets.h:24]: Btw can you elaborate on what "marking the `closedPromise` as handled explicitly" means? how do I do that?

Comment (jasnell) [reply src/workerd/api/sockets.h:24]: See the `markAsHandled()` method on `jsg::Promise` (https://github.com/cloudflare/workerd/blob/002cf0d582e9effd6bad46461c4ca4f242579563/src/workerd/jsg/promise.h#L232)

File: src/workerd/api/sockets.c++
Comment (jasnell) [reply src/workerd/api/sockets.c++:82]: Well, I'd allow the error to propagate to the `closed` promise but I would explicitly mark it handled using `markAsHandled` so that if someone isn't paying attention it's perfectly ok.

File: src/workerd/api/sockets.h
Comment (dom96) [reply src/workerd/api/sockets.h:24]: Okay, I added it where I construct the `PromiseResolverPair` (couldn't figure out how to access it through the `jsg::MemoizedIdentity`)

File: src/workerd/api/sockets.c++
Comment (dom96) [reply src/workerd/api/sockets.c++:82]: > Meanwhile it's not clear to me whether the closed promise should signal errors or just signal that the socket is now closed.

It's not clear to me either. Though I trust James' opinion here.

This is another case that I think we need to discuss in more depth. I want us to do so on this issue and many others, after discussion we may end up with another decision (especially if we decide to conform to [other specs](https://wicg.github.io/direct-sockets/)) so let's not worry too much about making it perfect in this PR.

**Review by kentonv — APPROVED** [review]

**Comment by jasnell** [comment]
@dom96 ... PR needs a rebase.

**Pull request merged**
